packages/services/src/risk-metrics.service.ts

```diff
@@ -0,0 +1,143 @@
+export interface AllRiskMetricsResult {
+  metrics: SelectPerpsRiskMetrics;
+}
+
+/**
+ * Orchestrator service for calculating all risk metrics atomically
+ * Coordinates CalmarRatioService and SortinoRatioService within a database transaction
+ * to ensure all-or-nothing persistence
+ */
+export class RiskMetricsService {
+  private calmarRatioService: CalmarRatioService;
+  private sortinoRatioService: SortinoRatioService;
+  private perpsRepo: PerpsRepository;
+  private competitionRepo: CompetitionRepository;
+  private db: Database;
+  private logger: Logger;
+
+  constructor(
+    calmarRatioService: CalmarRatioService,
+    sortinoRatioService: SortinoRatioService,
+    perpsRepo: PerpsRepository,
+    competitionRepo: CompetitionRepository,
+    db: Database,
+    logger: Logger,
+  ) {
+    this.calmarRatioService = calmarRatioService;
+    this.sortinoRatioService = sortinoRatioService;
+    this.perpsRepo = perpsRepo;
+    this.competitionRepo = competitionRepo;
+    this.db = db;
+    this.logger = logger;
+  }
+
+  /**
+   * Calculate and save all risk metrics (Calmar and Sortino) atomically
+   * Executes both calculations sequentially within a transaction to ensure
+   * consistent state - either both metrics are saved or neither
+   *
+   * @param agentId Agent ID
+   * @param competitionId Competition ID
+   * @returns Combined risk metrics after successful calculation and persistence
+   */
+  async calculateAndSaveAllRiskMetrics(
+    agentId: string,
+    competitionId: string,
+  ): Promise<AllRiskMetricsResult> {
+    try {
+      this.logger.info(
+        `[RiskMetrics] Starting atomic risk metrics calculation for agent ${agentId}`,
+      );
+
+      const result = await this.db.transaction(async (tx) => {
+        // Calculate both metrics (no longer saving individually)
+        const calmarMetrics =
+          await this.calmarRatioService.calculateCalmarRatio(
+            agentId,
+            competitionId,
+          );
+
+        const sortinoMetrics =
+          await this.sortinoRatioService.calculateSortinoRatio(
+            agentId,
+            competitionId,
+          );
+
+        // Combine metrics into single record
+        const combinedMetrics = {
+          ...calmarMetrics,
+          ...sortinoMetrics,
+          // Ensure we don't duplicate common fields
+          agentId,
+          competitionId,
+        };
+
+        // Single atomic save
+        const savedMetrics = await this.perpsRepo.upsertRiskMetrics(
+          combinedMetrics,
+          tx,
+        );
+
+        // Get latest portfolio snapshot for timestamp
+        const snapshots = await this.competitionRepo.getAgentPortfolioSnapshots(
+          competitionId,
+          agentId,
+          1, // Only need latest
+        );
+
+        const latestSnapshot = snapshots[0];
+
+        if (latestSnapshot) {
+          // Save time-series snapshot with all metrics
+          await this.perpsRepo.batchCreateRiskMetricsSnapshots(
```

Comment:
should we build the batch outside of this block, and then pass _all_ agents, not just a single agent ID? 

i.e., call `batchCreateRiskMetricsSnapshots` with an array of many values, which means the fn sig for `calculateAndSaveAllRiskMetrics` would need to change

Reply:
the current approach includes fault isolation, parallel processing, granular retry logic, clear logs per agent, and holds only one agent's data per worker. downside is more db round-trips + more connection overheat

if we were to move to a batch, approach, we'd gain the efficiency of fewer db round trips and true atomicity for all agents, but at the cost of all or nothing failure, no parallel processing, complex retry logic,  long-running db transactions, and the memory presh of holding all calculated data in memory until written to db.

i'm thinking that given the complexity of the risk metrics (esp once we start including others like sharpe, etc), the current approach is the right trade off - please let me know if you feel strongly otherwise